cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -0,0 +1,195 @@
+        try {
+          ((NonTransactionalHttpContentConsumer) delegate).onReceived(request.toByteBuffer(), new Transactional() {
+            @Override
+            public void execute(TxRunnable runnable) {
+              // Need to reset the context classloader to the system one before calling the
+              // actual transactional.execute(). The classloader will get switched again inside the
+              // transactional.execute() method before calling user TxRunnable.run() method
+              ClassLoader userClassLoader = ClassLoaders.setContextClassLoader(oldClassLoader);
+              try {
+                transactional.execute(runnable);
+              } finally {
+                ClassLoaders.setContextClassLoader(userClassLoader);
+              }
+            }
+          });
+        } finally {
+          ClassLoaders.setContextClassLoader(oldClassLoader);
+        }
+      } else {
+        transactional.execute(new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            delegate.onReceived(request.toByteBuffer());
+          }
+        });
+      }
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void finished(HttpResponder responder) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onFinish(BodyConsumerAdapter.this.responder);
+        }
+      });
+      BodyConsumerAdapter.this.responder.execute();
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void handleError(final Throwable cause) {
+    // When this method is called from netty-http, the response has already been sent, hence uses a no-op
+    // DelayedHttpServiceResponder for the onError call.
+    onError(cause, new DelayedHttpServiceResponder(responder) {
+      @Override
+      protected void doSend(int status, ByteBuffer content, String contentType,
+                            @Nullable Multimap<String, String> headers, boolean copy) {
+        // no-op
+      }
+
+      @Override
+      public void setTransactionFailureResponse(Throwable t) {
+        // no-op
+      }
+
+      @Override
+      public void execute() {
+        // no-op
+      }
+    });
+  }
+
+  /**
+   * Calls the {@link HttpContentConsumer#onError(HttpServiceResponder, Throwable)} method from a transaction.
+   */
+  private void onError(Throwable cause, final DelayedHttpServiceResponder responder) {
+    final Throwable failureCause = (cause instanceof WrappedException) ? cause.getCause() : cause;
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onError(responder, failureCause);
+        }
+      });
+    } catch (Throwable t) {
+      responder.setTransactionFailureResponse(t);
```

Comment:
So: if onReceived fails, we wrap the exception into a WrappedException and call this method. This again calls Transactional.execute(), so if that throws an exception, it is a WrappedException, and we respond with that? Seems not right.